src/pybind/mgr/k8sevents/module.py

```diff
@@ -0,0 +1,1029 @@
+# Integrate with the kubernetes events API. 
+# This module sends events to Kubernetes, and also captures/tracks all events
+# in the rook-ceph namespace so kubernetes activity like pod restarts,
+# imagepulls etc can be seen from within the ceph cluster itself.
+#
+# To interact with the events API, the mgr service to access needs to be 
+# granted additional permissions
+# e.g. kubectl -n rook-ceph edit clusterrole rook-ceph-mgr-cluster-rules
+#
+# These are the changes needed;
+# - apiGroups:
+#   - ""
+#   resources:
+#   - events
+#   verbs:
+#   - create
+#   - patch
+#   - list
+#   - get
+#   - watch
+
+# LogEntry class dynamically builds an object from a set of args,
+# disable the linter for the no-member error
+# pylint: disable=no-member
+
+
+import os
+import sys
+import time
+import json
+import threading
+
+from datetime import datetime, timedelta, tzinfo
+from urllib3.exceptions import MaxRetryError
+from collections import OrderedDict
+
+import rados
+from mgr_module import MgrModule
+from ceph_argparse import json_command, run_in_thread
+
+try:
+    import queue
+except ImportError:
+    # python 2.7.5
+    import Queue as queue
+finally:
+    # python 2.7.15 or python3
+    event_queue = queue.Queue()
+
+try:
+    from kubernetes import client, config, watch
+    from kubernetes.client.rest import ApiException
+except ImportError:
+    kubernetes_imported = False
+    client = None
+    config = None
+else:
+    kubernetes_imported = True
+    # load kubernetes config
+    config.load_incluster_config()
+
+DEFAULT_LOG_LEVEL = 'info'
+DEFAULT_HEARTBEAT_INTERVAL_SECS = 3600  # 1hr default
+DEFAULT_CONFIG_CHECK_SECS = 10
+
+ZERO = timedelta(0)
+
+
+def text_suffix(num):
+    """Define a text suffix based on a value i.e. turn host into hosts"""
+    return '' if num == 1 else 's'
+
+
+class UTC(tzinfo):
+    """UTC"""
+
+    def utcoffset(self, dt):
+        return ZERO
+
+    def tzname(self, dt):
+        return "UTC"
+
+    def dst(self, dt):
+        return ZERO
+
+class HealthCheck(object):
+    """Transform a healthcheck msg into it's component parts"""
+
+    def __init__(self, msg):
+
+        # msg looks like
+        # 2019-07-17 02:59:12.714672 mon.a [WRN] Health check failed: Reduced data availability: 100 pgs inactive (PG_AVAILABILITY)
```

Comment:
Good question. What do you have in mind?

Reply:
A Teuthology test would work, but only without a rook cluster.

@jecluis is there a way to get those messages before they are rendered into a string?

Reply:
this is still unresolved.

Reply:
@liewegas I don't think it is a good idea to parse these health messages in this mgr module. The code looks very brittle. Do you see any way to avoid this?

Reply:
The problem here is, you're introducing an unverified and undocumented dependency. Someone changing the log message will be unaware of the code here that tries to parse them. 

* One idea would be to move the code to parse the messages to a location outside of this module and make sure it is properly tested in a Teuthology test.
* Another idea would be to explicitly message this module to let it know of changes it should push to k8s. Meaning changing the C++ code and call this module. 
* Or, you could enhance https://github.com/sebastian-philipp/test-rook-orchestrator to include this module and add this to Sepia's Jenkins.

Reply:
The module emits events to k8s from ceph 'log' records based on the audit channel (json format), or the cluster channel. For records it sees in the cluster channel, it looks for "health check" in the message (this is defined in src/mon/Monitor.cc) - everything else is 'not worthy'. Monitor.cc processes "checks" objects so they're already in a specific format - which means anyone adding a healthcheck must adhere to the expected format...right? If I was processing freeform log messages, I'd agree with your comments. Doesn't this mean that the risk here isn't that someone adds a healthcheck that would be missed, but an k8sevents incompatibility if the health check subsystem in ceph itself is changed.

Reply:
> which means anyone adding a healthcheck must adhere to the expected format...right? 

The format of health checks can is here: `ceph -s -f json | jq .health`. The log messages are just that: log messages.

> If I was processing freeform log messages, I'd agree with your comments. Doesn't this mean that the risk here isn't that someone adds a healthcheck that would be missed, but an k8sevents incompatibility if the health check subsystem in ceph itself is changed.

Right, the health check subsystem was changed for example in #29422 . There could have been change that marked checks as muted in your log messages. Thankfully that was not the case. 

I'm not arguing against parsing log messages. I'm just saying that this module doesn't contain any protection against changes of the log messages. Things will break and we need to protect this module from getting broken without anyone noticing it.